### DISCUSSION TITLE
Requirements setup for fprime in a Docker container, 'Matching python package version/distribution not found' family of errors, and related fun stuff

### DISCUSSION THREAD

**Discussion by Anirban166** [nasa/fprime#1608] [Q&A]
In the process of trying to set up fprime in a docker container (to facilitate easy execution of a tool that is configured therein, which in turn requires me to install this FSW framework atop if I were to integrate both) running Linux, I got the 'Could not find a version that satisfies the requirement `package_name`==`semantic_version`' or 'No matching distribution found for `xyz_package`==`major.minor.patch`' type of errors when installing the required python packages via pip:
```sh
$ python3 -m pip install -U -r requirements.txt
Defaulting to user installation because normal site-packages is not writeable
Collecting aniso8601==9.0.1
  Downloading aniso8601-9.0.1-py2.py3-none-any.whl (52 kB)
     |################################| 52 kB 2.6 MB/s             
Collecting argcomplete==1.12.3
  Downloading argcomplete-1.12.3-py2.py3-none-any.whl (38 kB)
Collecting arrow==1.2.2
  Downloading arrow-1.2.2-py3-none-any.whl (64 kB)
     |################################| 64 kB 5.0 MB/s             
Collecting attrs==21.4.0
  Downloading attrs-21.4.0-py2.py3-none-any.whl (60 kB)
     |################################| 60 kB 5.8 MB/s             
Collecting binaryornot==0.4.4
  Downloading binaryornot-0.4.4-py2.py3-none-any.whl (9.0 kB)
Collecting Brotli==1.0.9
  Downloading Brotli-1.0.9-cp36-cp36m-manylinux1_x86_64.whl (357 kB)
     |################################| 357 kB 11.6 MB/s            
Collecting certifi==2022.6.15
  Downloading certifi-2022.6.15-py3-none-any.whl (160 kB)
     |################################| 160 kB 13.8 MB/s            
Collecting chardet==4.0.0
  Downloading chardet-4.0.0-py2.py3-none-any.whl (178 kB)
     |################################| 178 kB 7.8 MB/s            
Collecting charset-normalizer==2.0.12
  Downloading charset_normalizer-2.0.12-py3-none-any.whl (39 kB)
Collecting Cheetah3==3.2.6.post1
  Downloading Cheetah3-3.2.6.post1-cp36-cp36m-manylinux1_x86_64.whl (673 kB)
     |################################| 673 kB 6.9 MB/s            
Collecting click==7.1.2
  Downloading click-7.1.2-py2.py3-none-any.whl (82 kB)
     |################################| 82 kB 3.1 MB/s             
Collecting cookiecutter==1.7.3
  Downloading cookiecutter-1.7.3-py2.py3-none-any.whl (34 kB)
Collecting Flask==1.1.4
  Downloading Flask-1.1.4-py2.py3-none-any.whl (94 kB)
     |################################| 94 kB 6.3 MB/s             
Collecting Flask-Compress==1.12
  Downloading Flask_Compress-1.12-py3-none-any.whl (7.9 kB)
Collecting Flask-RESTful==0.3.9
  Downloading Flask_RESTful-0.3.9-py2.py3-none-any.whl (25 kB)
ERROR: Could not find a version that satisfies the requirement fprime-fpp==1.0.2 (from versions: 1.0.1, 1.0.2a8, 1.0.2a10, 1.0.2b1, 1.0.2b6, 1.0.2b10, 1.0.2b11, 1.0.2b12, 1.0.2b13)
ERROR: No matching distribution found for fprime-fpp==1.0.2
```
If I were to remove `fprime-fpp` from the requirements file (with the intention of manually installing it later) and retry, I halt at the very next (lexicographically) required package:
```sh
ERROR: Could not find a version that satisfies the requirement fprime-gds==3.1.1 (from versions: 1.5.5a1, 2.0, 3.0.0, 3.0.1, 3.0.2)
ERROR: No matching distribution found for fprime-gds==3.1.1
```
Repeating what I did above once again, I'm presented with the next one:
```sh
ERROR: Could not find a version that satisfies the requirement fprime-tools==3.1.0 (from versions: 1.5.4.dev0, 1.5.4.dev1, 1.5.5a1, 2.0.0, 2.0.1, 2.0.2, 3.0.0, 3.0.1)
ERROR: No matching distribution found for fprime-tools==3.1.0
```
And this goes on, although it's not that all packages continually throw this, and some down the line do, in fact, get installed (but it seems like the 'fprime' prefixed ones are among the chosen ones). 

This [PR](https://github.com/nasa/fprime/pull/1513) shines some light on this, but the solution itself doesn't inherently work for all cases, as things can get more complicated if Python, pip, and some interlinked dependencies are not configured well. (probably worth mentioning that I haven't faced any problems outside the container while following the regular setup)

**Comment by astroesteban** [comment]
Perhaps you are missing the `setuptools` packages?
```
python3 -m pip install -U --upgrade pip setuptools setuptools_sc
```

I also have a devcontainer repo: https://github.com/astroesteban/fprime-devcontainer

I use that regularly to develop for fprime inside of a Docker container with Visual Studio Code. You could probably take some of the stuff I have in there for your own purposes.

**Reply by Anirban166** [reply]
Nope, I tried your commands early on but like I mentioned in my post above, they didn't work (I still got the same version not found errors, with no changes at all, i.e. those two errors kept recurringly popping up for certain packages) in my case even when I had `setuptools` installed right (and following that setup from your PR, I remember that I still had to symlink it to my local path for `setuptools` to be recognized as back then I didn't execute as `sudo` at first; prior to that, I also ensured that separate possibly conflicting versions such as older `python-setuptools` were not existent).

The problem at its core was Python (and thus pip) not being upgraded to the main version - when you try to do so, it either doesn't upgrade (corrupted files or locked-in versions) or it doesn't point to the correct updated version (you gotta do this manually). I uninstalled and purged both python and pip installations, then had a fresh copy of Python 3.8.0 and pip 22.2.1 installed separately. I had to update my alternatives as `python3 --version` still returned 3.6.9 (what I had in my container from the start), and make it point to the latest installation. This is what I ran to configure after having 3.8 listed as an alternative:
```sh
sudo update-alternatives --config python3
```
That updated my Python3 and made the requirements installation through `sudo python3 -m pip install -r requirements.txt` run without any problems (some things still had to be fixed down the road, but this command executed successfully till the end and did not throw those errors for any package). And I didn't explicitly install `setuptools` in this current version of my container (I spun up a new one from my base image midway), but it still worked. I reckon that getting Python3.8 installed right would have brought that in (along with pip, although I installed that separately anyway). 

> I also have a devcontainer repo: https://github.com/astroesteban/fprime-devcontainer
>
> I use that regularly to develop for fprime inside of a Docker container with Visual Studio Code. You could probably take some of the stuff I have in there for your own purposes.

Thanks for sharing the container, but I cannot do it the other way around, i.e. I need to be able to set up fprime on top of a preconfigured container that has a testing tool (which comes integrated with a collection of fuzzers and symbolic engines) set up right, and not use a container that has fprime set up and needs to have the tool installed and configured along with its dependencies (I remember discussing the same with Michael yesterday, and he mentioned that the latter approach is not recommended - I'm certain that it would be more difficult to do so as well, even if the first way/approach was seen as feasible)

Anyways, I got this resolved yesterday so don't worry. And thanks for the help! 
I will post an answer for what worked in my case, with the steps I followed in my container in the process. Still have to recollect and organize my thoughts and notes from my log before doing the brain dump (gonna be a bit lengthy, thinking to include further steps such as CMake updation as well) soon (at some point today).

**Reply by astroesteban** [reply]
That's great to hear that you got that bug resolved. It's a rather interesting one really. I'm certain the community will benefit from the steps you took to resolve that 🙂

**Comment by Anirban166** [comment]
Before diving into the solution, it's worth noting that manual installation (without pip) of the error-invoking packages is not a good idea since the specific versions even when installed directly from PyPI for the correct version might interface with python or pip later on and cause issues (as one or both of them are redundant, and that's the root of this problem), apart from the fact that it can be cumbersome. The first step in itself while going that route would be to distinguish between what is there vs what's not, for which one could compare the currently installed pip packages with the ones from the requirements:
```sh
python3 -m pip list > installedPackages.txt
diff installedPackages.txt <fprime-checkout-path>/requirements.txt
```
Pretty sure there are better approaches to do the same, but I suspect it'll still be a bit of a pain.

### Solution
As previously mentioned in this discussion and as documented elsewhere on the internet to be the core solution to this class of errors, one would have to uninstall both python(3) and pip(3), apart from completely removing the python packages that are partially installed or corrupted, and then install fresh versions of them using the correct method(s). Unfortunately, there are a handful of ways to go about this, and picking the right one(s) might not be obvious at times. (at least for me it wasn't at the beginning, with my inexperience of dabbling with python packages)

By trial and error, a rule of thumb I learned and followed for the successful build was to use the correct package manager for what it's primarily intended to do, to not stick with one for all intents and purposes, and to definitely not overlap them. For e.g., if I were to introduce software such as python and pip, I'd use apt-get in Linux, and if I were to introduce python packages, I'd use pip. Mixing them (such as installing/updating python packages with both apt-get and pip) can lead to potential errors and misconfiguration.

For uninstalling, one could use these: (I used the second option for my current build for clearing the leftover config files, but the first might be enough too, depending upon how python-based stuff is set up in your container)
```sh
sudo apt-get remove python-pip python3-pip python python3
# assume the arguments above^ to be an ellipsis (...) henceforth
# better to use purge to remove the remnant config files:
sudo apt-get purge ...
# and possibly even better with this combination, if all dependencies need to be removed:
sudo apt-get --purge autoremove ...
```
The next step is to get them back afresh (at some point, I'm sure I used `--reinstall` instead of going with the remove-purge-install idiom, but if I remember correctly, that didn't quite do the same job as the latter approach). <br>
For (re)installing, I start with python3, going with a version that is acceptable: (fprime at present mandates 3.7+, so I picked the next minor number i.e. 8)
```sh
sudo apt-get install python3.8
```
While this does set up python3.8 right, the old version (3.6.9 in my case) was still the one being defaulted to when `python3` is used in any context (I did expect the `python3` removal and purging to have done a more thorough job, but this persisted for the moment, not sure why though), and running `python3 --version` returned `Python 3.6.9`. 

As such, one would need to set `python3` to point to the updated version, and the `update-alternatives` command helps with this directly. The arguments for this command with the `--install` option includes the symbolic link (get it from `which packageName`), name (`packageName`), and the priority (an integer, indicating the order of version selection). Emplace them accordingly and it will set things right: 
```sh
sudo update-alternatives --install /usr/bin/python3 python3 /usr/bin/python3.8 1
```
```sh
update-alternatives: using /usr/bin/python3.8 to provide /usr/bin/python3 (python3) in auto mode
```
Given that I set the priority to the topmost integer (in descending order, like 3.6.9 at 2 would be of lower priority in comparison), it defaulted to that version, and running `python3 --version` now returns `Python 3.8.0` as expected. Not only was this set, even the old version somehow got automatically removed, as when I ran `sudo update-alternatives --config python3`, I got:
```sh
There is only one alternative in link group python3 (providing /usr/bin/python3): /usr/bin/python3.8
Nothing to configure.
```
With Python3.8 set up, pip3 (or rather I guess I should start calling it `python3 -m pip`) got installed as well.

For Python2.x and pip, I can't trace back and see my log as I did it in a separate CLI session (using Docker Desktop) prior to the current one that I've been running (and I didn't make notes of it), but I think I used these commands: (pulled this up from a thread that I was looking up around that time)
```sh
sudo apt-get update && sudo apt-get -y install python python-pip
curl "https://bootstrap.pypa.io/get-pip.py" -o "get-pip.py"
python get-pip.py
```
Either way, getting python3 and its corresponding pip package installer should be good enough for our case here.

At the end of all this initial setup, I check their versions and the places where the executable binaries for them reside: (should be within recognized paths)
```sh
declare -a pystuff=("python" "pip" "python3" "pip3")
for i in "${pystuff[@]}"; do "$i" --version; which "$i"; done
```
Here's what they looked like for me prior to a successful pip install of the requirements, and at present:
```sh
Python 2.7.17
/usr/bin/python
pip 9.0.1 from /usr/lib/python2.7/dist-packages (python 2.7)
/usr/bin/pip
Python 3.8.0
/usr/bin/python3
pip 22.2.1 from /home/user/.local/lib/python3.8/site-packages/pip (python 3.8)
/usr/local/bin/pip3
```
I proceeded to install the requirements and they *finally* worked for all the packages without a problem!
```sh
sudo python3 -m pip install -r requirements.txt
```
Note that using `sudo` here would install packages at their intended site as those directories become writeable, as opposed to them being restricted during the ordinary user installation.

Again, feel free to exercise `sudo` even though that is one of the things suggested to not be done while using pip when administrative permissions are not a mandatory thing or when the absence of it doesn't issue a roadblock. Stuff is being run on a docker container, so there is nothing to be worried about. Moreover, while installing `fprime-tools` or the `fprime-*` packages, it's beneficial to run it with `sudo` as this allows pip to directly emplace these tools/packages to <home directory>/.local/bin (avoiding the need to be in a virtual environment or to explicitly add that directory to the path), which in turn avoids `fprime-util` from being an unknown command. Also, `fprime-util purge` can be used to cleanse or remove the build caches in a directory (and this actually helped to get a python dependency to be recognized when `fprime-util check` was not having it to be. Reinstalling via `python3 -m pip uninstall <packageName> && sudo python3 -m pip install <packageName>` was tried but it appears that the package was in fact correctly installed, and the problem was with the build cache entries), which tends to resolve errors for the aforementioned point. (Thanks to @LeStarch for sharing this information!)

Speaking of cache also reminds me that redundant pip data can also lead to problems, such as during the deserialization of cached package entries:
```sh
Collecting aniso8601==9.0.1 (from -r fprime/requirements.txt (line 1))
  Cache entry deserialization failed, entry ignored
  Using cached https://files.pythonhosted.org/packages/e3/04/e97c12dc034791d7b504860acfcdd2963fa21ae61eaca1c9d31245f812c3/aniso8601-9.0.1-py2.py3-none-any.whl
Collecting argcomplete==1.12.3 (from -r fprime/requirements.txt (line 2))
  Cache entry deserialization failed, entry ignored
  Using cached https://files.pythonhosted.org/packages/b7/9e/9dc74d330c07866d72f62d553fe8bdbe32786ff247a14e68b5659963e6bd/argcomplete-1.12.3-py2.py3-none-any.whl
Collecting arrow==1.2.2 (from -r fprime/requirements.txt (line 3))
  Cache entry deserialization failed, entry ignored
...
```
Given that it's an issue with the pip cache entirely, clearing or removing it (`rm -rf ~/.cache/pip`) would do the job without any problems. One reason for this to occur as I suspect would be when the cache gets disabled if a directory isn't set to be writeable by the current user (in which case, it might need to be purged).

Now my goal with this container was to experiment with the files that contained STest-based unit tests wrapped under GoogleTest macros, so as to run them using `fprime-util check` (which internally invokes GTest). 
For this, one needs to generate and build the UTs before proceeding to check them. In order to achieve this, yet another requirement had to be upgraded in my container, which was none other than CMake (the only remaining one with a version specification other than Python3) that was under 3.16 (or the version currently specified in the <sup>1</sup>[requirements](https://github.com/nasa/fprime/blob/devel/docs/INSTALL.md#requirements)) back then. I upgraded it to the latest version (3.24.0 atm).

I think it might be helpful to share the steps for future reference if anyone else is trying to downgrade/update CMake (I see one [discussion](https://github.com/nasa/fprime/discussions/1073) with relevancy), so here's what I would suggest based on what I did:
- Delete the current version (I used `sudo apt remove cmake`).
- Download the applicable shell script for your architecture and operating system from the [official website](https://cmake.org/download/).
Here's how I usually check platform-specific details inside Linux containers: (the distribution and version aren't required to be known to figure out the CMake version though, just being extra)
```sh
arch && cat /etc/os-release | head -2
```
```sh
x86_64
NAME="Ubuntu"
VERSION="18.04.2 LTS (Bionic Beaver)"
```
Once you figured the one relevant for your arch and OS, download it to your computer (you could probably download it inside the Docker container through an URL directly, but I was lazy enough to avoid that route and instead just use the browser in my host machine to download the script, with the intention of copying it over to my container afterward). <br>
I like to set up a short alias (temporary, can `unset` later) for the name of the downloaded file (extension included) for use in the next steps (especially for use inside the container), given that its a bit lengthy:
```sh
cmscript=cmake-3.24.0-rc5-linux-x86_64.sh # identify your version and replace accordingly!
```
- Now get the ID of your container (use `docker ps` for a list of running ones), then copy the script over there from your host machine: 
```sh
sudo docker cp $cmscript <containerID>:/usr/share
```
- I collect in '/usr/share/', but feel free to copy/`cp` or move/`mv` the file to anywhere else if preferred:
```sh
sudo mv /usr/share/$cmscript /<directory>/
# for e.g., <directory> can be usr, usr/local, opt, etc.
```
- Grant the script the permission to execute, and run it: (I'm also skipping the license to avoid the scrolling plus double 'y' presses for mutual agreement, excluding subdirectories, and setting the prefix to the default location to reconfigure environment variables to have an installation therein)
```sh
sudo chmod +x /<directory>/$cmscript
sudo bash /<directory>/$cmscript --skip-license --exclude-subdir --prefix=/usr/local
```
- If a separate path is required for accessing the binary, create a symbolic link thereto:
```sh
sudo ln -s /usr/local/bin/cmake <alternatePath>
```
- *Finally*, test using `cmake --version` to confirm:
```
cmake version 3.24.0-rc5

CMake suite maintained and supported by Kitware (kitware.com/cmake).
```
<details>
<summary>Optional, but feel free to do a final version and executable location check for all the <sup>1</sup>requirements at this point, just to reassure oneself</summary>

```sh
declare -a allreqs=("cmake" "git" "gcc" "${pystuff[@]}")
for i in "${allreqs[@]}"; do "$i" --version; which "$i"; done
```
</details> 

If a container running OS X is to be used (I couldn't due to my tool's limited cross-platform support), I suspect this wouldn't be this complex with Homebrew. Just use `brew uninstall cmake` and reinstall using the right bottle. If `cmake --version` isn't showing up, it could be the case that the homebrew path hasn't been registered in your shell environment (could be in a regular session, or the integrated terminal invoked by an IDE such as VS Code if the default settings are not carried over), for which you can add that to the 'profile' suffixed dotfile that your shell uses for environment configuration, like for the default shell `zsh` it would be `echo 'eval "$(/opt/homebrew/bin/brew shellenv)"' >> ~/.zprofile`.
